File: python/algo/202512/algo3693.py
from cmath import inf
from functools import cache
from typing import List

# 动态规划 - 爬楼梯
class Solution:
    def climbStairs(self, n: int, costs: List[int]) -> int:
        # 自行解答
        @cache
        def dfs(i: int) -> int:
            if i == 0: return 0
            if i < 0: return inf
            
            return costs[i-1] + min(dfs(i-1) + 1, dfs(i-2) + 4 ,dfs(i-3) + 9)
        return dfs(n)

File: cplus/leetcode/202512/algo3693.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 - 爬楼梯
class Solution {
public:
    int climbStairs(int n, vector<int>& costs) {
        // 自行解答
        vector<int> dp(n+1);
        for (int i=1; i<=n; ++i) {
            int mn = INT_MAX;
            if (i-1>=0) mn = min(mn, dp[i-1] + 1);
            if (i-2>=0) mn = min(mn, dp[i-2] + 4);
            if (i-3>=0) mn = min(mn, dp[i-3] + 9);
            dp[i] = mn + costs[i];
        }

        return dp[n];
    }
};